pygbm/splitting.py

```diff
@@ -110,95 +96,212 @@ def find_node_split(self, sample_indices):
+            histogram = _build_histogram_root(
+                n_bins, binned_feature, ordered_gradients,
+                ordered_hessians)
     else:
         if constant_hessian:
-            histogram = _build_gc_histogram_unrolled(
+            histogram = _build_histogram_no_hessian(
                 n_bins, sample_indices, binned_feature, ordered_gradients)
         else:
-            hessian_parent = ordered_hessians.sum()
-            histogram = _build_ghc_histogram_unrolled(
+            histogram = _build_histogram(
                 n_bins, sample_indices, binned_feature, ordered_gradients,
                 ordered_hessians)
+
+    return _find_best_bin_to_split_helper(
+        feature_idx, n_bins, histogram, min_hessian_to_split,
+        l2_regularization, gradient, hessian,
+        constant_hessian, constant_hessian_value
+    )
+
+
+@njit(locals={'histogram': typeof(HISTOGRAM_DTYPE)[:],
+              'gradient': float32,
+              'hessian': float32},
+      fastmath=True)
+def _find_histogram_split_subtraction(feature_idx, binned_feature, n_bins,
+                                      sample_indices, l2_regularization,
+                                      min_hessian_to_split, parent_histograms,
+                                      sibling_histograms, constant_hessian,
+                                      constant_hessian_value, gradient,
+                                      hessian):
+    """Compute the histogram for a given feature and return the best bin to
+    split on. Uses hist(parent) = hist(left) + hist(right)"""
+
+    histogram = _subtract_histograms(
+        n_bins, parent_histograms[feature_idx],
+        sibling_histograms[feature_idx]
+    )
+
+    return _find_best_bin_to_split_helper(
+        feature_idx, n_bins, histogram, min_hessian_to_split,
+        l2_regularization, gradient, hessian,
+        constant_hessian, constant_hessian_value
+    )
+
+
+@njit(locals={'gradient_left': float32, 'hessian_left': float32},
+      fastmath=True)
+def _find_best_bin_to_split_helper(feature_idx, n_bins, histogram,
+                                   min_hessian_to_split, l2_regularization,
+                                   gradient, hessian, constant_hessian,
+                                   constant_hessian_value):
+    """For a given feature and the corresponding histogram, find best bin to
+    split on and return the corresponding SplitInfo"""
+    # Allocate the structure for the best split information. It can be
+    # returned as such (with a negative gain) if the min_hessian_to_split
+    # condition is not satisfied. Such invalid splits are later discarded by
+    # the TreeGrower.
+    best_split = SplitInfo(-1., 0, 0, 0., 0., 0., 0.)
+
     gradient_left, hessian_left = 0., 0.
-    for bin_idx in range(histogram.shape[0]):
+    for bin_idx in range(n_bins):
         gradient_left += histogram[bin_idx]['sum_gradients']
```

Comment:
I have a weird feeling about this `+=` sign. The histogram is already a cumulative sum as far as I understand, so why do we want a `+=` sign here? I feel like it should be `=`.

Maybe this is related to my previous comment: https://github.com/ogrisel/pygbm/pull/14#issuecomment-430034908

Reply:
`histogram[bin_idx]['sum_gradients']` is the sum of the gradients w.r.t. the samples that fell precisely in `bin_idx` (as done in `_build_histogram`. `gradient_left` is the sum of gradients w.r.t. the samples of any bin <= `bin_idx`.

So I believe the current code with `+=` is correct.

Reply:
Ok thanks, I thought maybe `histogram[bin_idx]['sum_gradients']` was already the sum of gradients w.r.t. the samples of any bin <= bin_idx.

So yes the current code must be correct, but then the gradient at the current node should indeed be `sum(histogram[:]['sum_gradients'])`. Something fishy is going on (at least in my head ^^), I'll try to add some tests and identify a potential mistake I made.